crypto/kem/README.md

```diff
@@ -0,0 +1,159 @@
+```
+
+The `ctx` variable above is a pointer to a “context” object of type `EVP_PKEY_CTX`. The `pkey` variable listed as an argument of `EVP_PKEY_keygen` function is a pointer to an object of type `EVP_PKEY`. The `ciphertext` and `shared_secret` are byte arrays of length `ciphertext_len` and `shared_secret_len`, respectively.
+
+**The context.** The customer needs to create the context. Depending on the use case, this can be done in two ways:
+
+```
+1. EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL);
+   EVP_PKEY_CTX_kem_set_params(ctx, NID_KYBER512);
+```
+
+This creates a fresh context of type `EVP_PKEY_KEM` and sets the specific KEM parameters (Kyber512 in this example). The context is now ready for key generation (`EVP_PKEY_keygen`). However, the context created in this way doesn’t have an associated key (`EVP_PKEY`), so obviously, we can’t encapsulate/decapsulate with it. Therefore, this is useful for key generation only, i.e. before we have a key.
+
+The second way to create a context is with an already existing key (`EVP_PKEY`):
+
+```
+2. EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);
+```
+
+This creates a fresh context of type `EVP_PKEY_KEM`, sets the specific KEM parameters defined in the `pkey` object itself, and associates `pkey` to the context. The context is now ready to encapsulate/decapsulate because it contains the encapsulation/decapsulation key. The key is usually either generated by ourselves via `EVP_PKEY_keygen`, or we received the public key of the other party as a raw byte array and created the key as shown below.
+
+**EVP_PKEY from/to raw data.** The customer sometimes needs to store/load the key-pair (`public_key, secret_key`) to/from memory. The customer certainly needs to be able to send/receive the `public_key` as part of the protocol. Therefore, we define the following methods:
+
+```
+1. Create EVP_PKEY object from raw data:
+
+   EVP_PKEY *EVP_PKEY_kem_new_raw_key(int nid,
+                                      const uint8_t *in_public, size_t len_public,
+                                      const uint8_t *in_secret, size_t len_secret);                             
+   EVP_PKEY *EVP_PKEY_kem_new_raw_public_key(int nid, const uint8_t *in, size_t len);
+   EVP_PKEY *EVP_PKEY_kem_new_raw_secret_key(int nid, const uint8_t *in, size_t len);
+```
+
+and
+
+```
+2. Extract raw data from EVP_PKEY object:
+
+   OPENSSL_EXPORT int EVP_PKEY_get_raw_private_key(const EVP_PKEY *pkey,
+                                                   uint8_t *out, size_t *out_len);
+   OPENSSL_EXPORT int EVP_PKEY_get_raw_public_key(const EVP_PKEY *pkey,
+                                                  uint8_t *out, size_t *out_len);
+```
+
+**KEM parameters size.** The customer needs to know the size of they keys, the ciphertext, and the shared secret. All the sizes can be retrieved by calling the above listed APIs with `NULL` parameters in place of the actual output pointers. For example, calling:
```

Comment:
not really, because the exact location of the declaration/definition can change and we'd always have to remember ti update it here. I'm afraid it'll be out of sync quickly.